A36-xds-for-servers.md

```diff
@@ -0,0 +1,324 @@
+certificates. While the user-facing API is discussed here, the precise
+implementation details are covered in gRFC A29. The user is free to pass other
+ServerCredentials types to XdsServer and they will be honored. Note that RBAC is
+authz, not authn, so it would be enabled by using the XdsServer, even without
+XdsServerCredentials.
+
+When starting the XdsServer, it will start communication with xDS to receive
+initial configuration. `bind()` and `listen()` may happen immediately, but
+`accept()`ing on the port must be delayed until the initial configuration is
+received. `listen()` should be delayed if possible, but commonly needs to happen
+immediately after `bind()` for implementation-specific reasons. If initial
+configuration fails, the server will continue delaying and server startup will
+stall until it succeeds. Each language will provide an XdsServer-specific API to
+inform the application of initial communication failures. If the user does not
+configure the API, gRPC should log the errors at a default-visible log level. If
+the xds bootstrap is missing or invalid, implementations would ideally fail
+server startup, but a permanent hang is also acceptable.
+
+The XdsServer does not have to wait until server credentials (e.g., TLS certs)
+are available before accepting connections; since XdsServerCredentials might not
+be used, the server is free to lazily load credentials. However, the XdsServer
+should keep the credentials cache fresh and up-to-date after that initial
+lazy-loading, as it is clear at that point that XdsServerCredentials are being
+used.
+
+The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+```json
+{
+  "grpc_server_resource_name_id": "example/resource",
+  // ...
+}
+```
+
+XdsServer will use the normal `XdsClient` to communicate with the xDS server.
+The XdsServer will pass the listening address (commonly using a wildcard IP
+address, like `::` or `0.0.0.0`) to a new XdsClient API to start a watch for a
+`envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
+value of `grpc_server_resource_name_id` from the bootstrap,
+`?udpa.resource.listening_address=`, and the listening address to construct the
+resource name. For example, with an address of `[::]:80` and the example
+bootstrap field above, the resource name would be
+`example/resource?udpa.resource.listening_address=[::]:80`. Note that the `[]`
+is _not_ percent encoded, as this string is not a URI.
+
+The xDS-returned Listener must have an [`address`][Listener.address] that
+matches the listening address provided. The Listener's `address` would be a TCP
+`SocketAddress` with matching `address` and `port_value`. The XdsClient must
+NACK the resource if the address does not match.
+
+Although `Filter`s will not be observed initially, the `FilterChain` contains
+data that may be used. When looking for a FilterChain, the standard matching
+logic must be used. Each `filter_chain_match` of the repeated
```

Comment:
Filter and related objects (FilterChain and FIlterChainMatch) are very Envoy specific. For gRPC we have already modified the listener query to supply the resource name (`example/resource?udpa.resource.listening_address=[::]:80` in the above example) which Envoy doesn't use. We can also modify the response specification to say that the response should not contain any filterChainMatch since they don't apply that well to gRPC. The rationale is that the control plane will have to make changes to support gRPC specific listener query so they will also make changes to send us responses without any filter data.